### ISSUE TITLE
ios_config crashes when there's no "dest" parameter

### ISSUE THREAD

**Issue by ipspace** [ansible/ansible-modules-core#4419]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ios_config
##### ANSIBLE VERSION

```
ansible-playbook 2.2.0
  config file = /vagrant/Ansible/Networking/ansible.cfg
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

Ansible: Ubuntu
Remote: Cisco IOS 15.5
##### SUMMARY

ios_config.py crashes with KeyError when the **dest** parameter is not specified
##### STEPS TO REPRODUCE

```

---
- hosts: r1.lab.local
  tasks:
  - name: "Configure SNMP on IOS devices"
    ios_config:
      provider: "{{cli}}"
      save_config: no
      lines:
        - "snmp-server community {{snmp_community}} RO"
        - "snmp-server contact {{snmp_contact}}"
        - "snmp-server host {{snmp_host}} {{snmp_community}}"
```

##### ACTUAL RESULTS

```
TASK [Configure SNMP on IOS devices] *******************************************
task path: /vagrant/Ansible/Networking/config-simple.yml:4
Using module file /usr/local/lib/python2.7/dist-packages/ansible-2.2.0-py2.7.egg/ansible/modules/core/network/ios/ios_config.py
<r1.lab.local> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<r1.lab.local> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1471070597.68-36067224165396 `" && echo ansible-tmp-1471070597.68-36067224165396="` echo $HOME/.ansible/tmp/ansible-tmp-1471070597.68-36067224165396 `" ) && sleep 0'
<r1.lab.local> PUT /tmp/tmpkiEnTr TO /home/vagrant/.ansible/tmp/ansible-tmp-1471070597.68-36067224165396/ios_config.py
<r1.lab.local> EXEC /bin/sh -c 'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1471070597.68-36067224165396/ /home/vagrant/.ansible/tmp/ansible-tmp-1471070597.68-36067224165396/ios_config.py && sleep 0'
<r1.lab.local> EXEC /bin/sh -c '/usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1471070597.68-36067224165396/ios_config.py; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1471070597.68-36067224165396/" > /dev/null 2>&1 && sleep 0'
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_Uq2ayW/ansible_module_ios_config.py", line 295, in <module>
    main()
  File "/tmp/ansible_Uq2ayW/ansible_module_ios_config.py", line 256, in main
    invoke('check_args', module, warnings)
  File "/tmp/ansible_Uq2ayW/ansible_module_ios_config.py", line 202, in invoke
    return func(*args, **kwargs)
  File "/tmp/ansible_Uq2ayW/ansible_module_ios_config.py", line 210, in check_args
    if module.params['dest'] and module.params['save_config'] is True:
KeyError: 'dest'

fatal: [r1.lab.local]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "lines": [
                "snmp-server community cisco RO",
                "snmp-server contact admin@lab.local",
                "snmp-server host 172.16.1.12 cisco"
            ],
            "provider": {
                "host": "r1.lab.local",
                "password": "cisco",
                "transport": "cli",
                "username": "cisco"
            },
            "save_config": false
        },
        "module_name": "ios_config"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_Uq2ayW/ansible_module_ios_config.py\", line 295, in <module>\n    main()\n  File \"/tmp/ansible_Uq2ayW/ansible_module_ios_config.py\", line 256, in main\n    invoke('check_args', module, warnings)\n  File \"/tmp/ansible_Uq2ayW/ansible_module_ios_config.py\", line 202, in invoke\n    return func(*args, **kwargs)\n  File \"/tmp/ansible_Uq2ayW/ansible_module_ios_config.py\", line 210, in check_args\n    if module.params['dest'] and module.params['save_config'] is True:\nKeyError: 'dest'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "parsed": false
}
```

**Comment by ipspace** [comment]
Suggest changing line 210 in ios_config.py to

```
    if module.params.get('dest') and module.params['save_config'] is True:
```

**Comment by ansibot** [comment]
@privateip, @gundalow, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
Hi,
Thanks for the bug report and suggested fix.

Doesn't appear to have been fixed in `devel` yet, so fix still needed.

**Comment by gundalow** [comment]
@ipspace Would you like to raise a PR to fix this as you have a test machine?

**Comment by ansibot** [comment]
@privateip, @gundalow, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ipspace** [comment]
Looks like that line is gone from the latest sources. Planning on redoing the tests tomorrow.

**Comment by gundalow** [comment]
Thanks :)

**Comment by ipspace** [comment]
Now it crashes in a totally different way ;)

Here's what I downloaded from Github this morning

```
ansible-playbook --version
ansible-playbook 2.2.0 (devel 03132041fb) last updated 2016/08/31 08:50:56 (GMT +000)
  lib/ansible/modules/core: (detached HEAD 5310bab12f) last updated 2016/08/31 08:51:16 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 2ef4a34eee) last updated 2016/08/31 08:51:33 (GMT +000)
  config file = /vagrant/Ansible/Networking/ansible.cfg
  configured module search path = Default w/o overrides
```

Here is the playbook:

```
---
- hosts: ios
  tasks:
  - ios_config: provider="{{cli}}" commands="no ip access-list extended AllowedTraffic"
```

And here is the crash dump:

```
TASK [ios_config] **************************************************************
task path: /vagrant/Ansible/Networking/config-acl-remove.yml:9
Using module file /home/vagrant/ansible/lib/ansible/modules/core/network/ios/ios_config.py
<r1.lab.local> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<r1.lab.local> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1472633929.36-16226392782709 `" && echo ansible-tmp-1472633929.36-16226392782709="` echo $HOME/.ansible/tmp/ansible-tmp-1472633929.36-16226392782709 `" ) && sleep 0'
<r1.lab.local> PUT /tmp/tmpz1tBfb TO /home/vagrant/.ansible/tmp/ansible-tmp-1472633929.36-16226392782709/ios_config.py
<r1.lab.local> EXEC /bin/sh -c 'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1472633929.36-16226392782709/ /home/vagrant/.ansible/tmp/ansible-tmp-1472633929.36-16226392782709/ios_config.py && sleep 0'
<r1.lab.local> EXEC /bin/sh -c '/usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1472633929.36-16226392782709/ios_config.py; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1472633929.36-16226392782709/" > /dev/null 2>&1 && sleep 0'
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_kRUKI_/ansible_module_ios_config.py", line 409, in <module>
    main()
  File "/tmp/ansible_kRUKI_/ansible_module_ios_config.py", line 399, in main
    invoke(state, module, result)
  File "/tmp/ansible_kRUKI_/ansible_module_ios_config.py", line 253, in invoke
    return func(*args, **kwargs)
  File "/tmp/ansible_kRUKI_/ansible_module_ios_config.py", line 331, in present
    backup_config(module)
  File "/tmp/ansible_kRUKI_/ansible_module_ios_config.py", line 299, in backup_config
    module.cli(cmd)
  File "/tmp/ansible_kRUKI_/ansible_modlib.zip/ansible/module_utils/netcli.py", line 68, in __call__
TypeError: 'Command' object is not iterable

fatal: [r1.lab.local]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "commands": "no ip access-list extended AllowedTraffic",
            "provider": {
                "host": "r1.lab.local",
                "password": "cisco",
                "transport": "cli",
                "username": "cisco"
            }
        },
        "module_name": "ios_config"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_kRUKI_/ansible_module_ios_config.py\", line 409, in <module>\n    main()\n  File \"/tmp/ansible_kRUKI_/ansible_module_ios_config.py\", line 399, in main\n    invoke(state, module, result)\n  File \"/tmp/ansible_kRUKI_/ansible_module_ios_config.py\", line 253, in invoke\n    return func(*args, **kwargs)\n  File \"/tmp/ansible_kRUKI_/ansible_module_ios_config.py\", line 331, in present\n    backup_config(module)\n  File \"/tmp/ansible_kRUKI_/ansible_module_ios_config.py\", line 299, in backup_config\n    module.cli(cmd)\n  File \"/tmp/ansible_kRUKI_/ansible_modlib.zip/ansible/module_utils/netcli.py\", line 68, in __call__\nTypeError: 'Command' object is not iterable\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE"
}
```

I fixed the error in line 299 in ios_config.py - changed it to

```
    module.cli([cmd])
```

Which made ios_config.py module to execute all commands on the router (including "exit") but then it hangs. Not sure I'd be able to debug that one...

**Comment by gundalow** [comment]
@ipspace Thanks for taking the time to retest this.

It's likely you've hit some of the refactoring that's going on as part of the preparation for Ansible 2.2.

I'll make this bug as requiring fixing for the 2.2.0 release so we ensure we get this tested.

**Comment by privateip** [comment]
@ipspace this should now be fixed.  will leave this issue open until we hear back that it tested okay

**Comment by ipspace** [comment]
This one is fixed (successfully removed and created an ACL), however there's another totally unrelated crash. Opening a new issue.

**Comment by gundalow** [comment]
@ipspace Thanks for confirming, if the other issue is in networking please add `@gundalow` and I'll have a look
